subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/repositories/DefaultRepositoryContentDescriptor.java

```diff
@@ -291,7 +307,11 @@ public boolean matches(ModuleIdentifier id) {
         public boolean matches(ModuleComponentIdentifier id) {
             return group.equals(id.getGroup())
                     && (module == null || module.equals(id.getModule()))
-                    && (version == null || version.equals(id.getVersion()));
+                    && (version == null || version.equals(id.getVersion()) || getVersionSelector().accept(id.getVersion()));
+        }
+
+        private VersionSelector getVersionSelector() {
+            return versionSelectors.computeIfAbsent(version, s -> versionSelectorScheme.parseSelector(version));
```

Comment:
I would initialize the `versionSelector` in the constructor. Otherwise you end up with a map lookup for every repository request.

Reply:
Good suggestion, done